src/components/Radio/Radio.tsx

```diff
@@ -1,66 +1,157 @@
-import { Extendable, ReactChildren } from '../../../types/utils'
+import { ComponentEventHandler, Extendable, ReactChildren } from '../../../types/utils'
 import { ComponentVariablesInput, IComponentPartStylesInput } from '../../../types/theme'
+import Icon from '../Icon/Icon'
 
 export interface IRadioProps {
   as?: any
+  checked?: boolean
   children?: ReactChildren
   className?: string
+  disabled?: boolean
   label?: string
+  name?: string
+  onChange?: ComponentEventHandler<IRadioProps>
   type?: string
   styles?: IComponentPartStylesInput
+  value?: string | number
   variables?: ComponentVariablesInput
 }
 
 /**
  * @accessibility
  * This is shown at the top.
  */
-class Radio extends UIComponent<Extendable<IRadioProps>, any> {
+class Radio extends AutoControlledComponent<Extendable<IRadioProps>, any> {
+  static create: Function
+
   static displayName = 'Radio'
 
   static className = 'ui-radio'
 
   static propTypes = {
     as: customPropTypes.as,
 
+    /** Whether or not radio is checked. */
+    checked: PropTypes.bool,
+
     /** Child content * */
     children: PropTypes.node,
 
     /** Additional classes. */
     className: PropTypes.string,
 
+    /** Initial checked value. */
+    defaultChecked: PropTypes.bool,
+
+    /** A radio can appear disabled and be unable to change states. */
+    disabled: PropTypes.bool,
+
     /** The label of the radio input. */
     label: PropTypes.string,
 
+    /** The HTML input name. */
+    name: PropTypes.string,
+
+    /**
+     * Called after radio is clicked.
+     * @param {SyntheticEvent} event - React's original SyntheticEvent.
+     * @param {object} data - All props.
+     */
+    onChange: PropTypes.func,
+
     /** Custom styles to be applied for component. */
     styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
 
     /** The HTML input type. */
     type: PropTypes.string,
 
+    /** The HTML input value. */
+    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+
     /** Custom variables to be applied for component. */
     variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
   }
 
-  static handledProps = ['as', 'children', 'className', 'label', 'styles', 'type', 'variables']
+  static handledProps = [
+    'as',
+    'checked',
+    'children',
+    'className',
+    'defaultChecked',
+    'disabled',
+    'label',
+    'name',
+    'onChange',
+    'styles',
+    'type',
+    'value',
+    'variables',
+  ]
 
   static defaultProps = {
     as: 'div',
     type: 'radio',
   }
 
-  renderComponent({ ElementType, classes, rest, styles }) {
-    const { type, label } = this.props
+  static autoControlledProps = ['checked']
+
+  private handleChange = (e: React.SyntheticEvent) => {
+    const { onChange, disabled } = this.props
+    const { checked } = this.state
+
+    if (disabled) {
+      e.preventDefault()
+      return
+    }
+
+    if (onChange) {
+      onChange(e, { ...this.props, checked: !checked })
```

Comment:
worrying about the fact that this callback method is invoked before state change will be applied to component - in case of potential error in the logic provided by the callback state change won't be applied

Reply:
Agreed, will move it before invoking the callback.